tracer/src/Datadog.Trace/Agent/StatsAggregator.cs

```diff
@@ -0,0 +1,197 @@
+        {
+            return settings.TracerStatsEnabled ? new StatsAggregator(api, settings, TimeSpan.FromSeconds(10)) : new NullStatsAggregator();
+        }
+
+        public Task DisposeAsync()
+        {
+            _processExit.TrySetResult(true);
+            return _flushTask;
+        }
+
+        public void Add(params Span[] spans)
+        {
+            AddRange(spans, 0, spans.Length);
+        }
+
+        public void AddRange(Span[] spans, int offset, int count)
+        {
+            // Contention around this lock is expected to be very small:
+            // AddRange is called from the serialization thread, and concurrent serialization
+            // of traces is a rare corner-case (happening only during shutdown).
+            // The Flush thread only acquires the lock long enough to swap the metrics buffer.
+            lock (_buffers)
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    AddToBuffer(spans[offset + i]);
+                }
+            }
+        }
+
+        internal async Task Flush()
+        {
+            // Use a do/while loop to still flush once if _processExit is already completed (this makes testing easier)
+            do
+            {
+                await Task.WhenAny(_processExit.Task, Task.Delay(_duration)).ConfigureAwait(false);
+
+                var buffer = CurrentBuffer;
+
+                lock (_buffers)
+                {
+                    _currentBuffer = (_currentBuffer + 1) % BufferCount;
+                }
+
+                if (buffer.Buckets.Count > 0)
+                {
+                    // Push the metrics
+                    await _api.SendStatsAsync(buffer, _duration.ToNanoseconds()).ConfigureAwait(false);
+
+                    buffer.Reset();
+                }
+            }
+            while (!_processExit.Task.IsCompleted);
+        }
+
+        /// <summary>
+        /// Converts a nanosec timestamp into a float nanosecond timestamp truncated to a fixed precision
+        /// </summary>
+        /// <param name="ns">Timestamp to convert</param>
+        /// <returns>Timestamp with truncated precision</returns>
+        private static double ConvertTimestamp(long ns)
+        {
+            // 10 bits precision (any value will be +/- 1/1024)
+            const long roundMask = 1 << 10;
+
+            int shift = 0;
+
+            while (ns > roundMask)
+            {
+                ns >>= 1;
+                shift++;
+            }
+
+            return ns << shift;
+        }
+
+        private static StatsAggregationKey BuildKey(Span span)
+        {
+            var rawHttpStatusCode = span.GetTag(Tags.HttpStatusCode);
+
+            if (rawHttpStatusCode == null || !int.TryParse(rawHttpStatusCode, out var httpStatusCode))
+            {
+                httpStatusCode = 0;
+            }
+
+            return new StatsAggregationKey(
```

Comment:
Another dimension we need to capture is `bool synthetics`, saying if the span came from a synthetic test or not

Reply:
Added in 4924a44116961e50deaeefc98bc98eb19d003bb5